l2geth/core/state_transition.go

```diff
@@ -185,26 +206,44 @@ func (st *StateTransition) useGas(amount uint64) error {
 }
 
 func (st *StateTransition) buyGas() error {
+	// max value
 	mgval := new(big.Int).Mul(new(big.Int).SetUint64(st.msg.Gas()), st.gasPrice)
+	tonval := new(big.Int)
+
 	if rcfg.UsingOVM {
 		// Only charge the L1 fee for QueueOrigin sequencer transactions
 		if st.msg.QueueOrigin() == types.QueueOriginSequencer {
-			mgval = mgval.Add(mgval, st.l1Fee)
+			if st.isFeeTokenUpdate == true {
+				mgval = mgval.Add(mgval, st.l1TonFee)
+			} else {
+				mgval = mgval.Add(mgval, st.l1Fee)
+			}
 			if st.msg.CheckNonce() {
-				log.Debug("Adding L1 fee", "l1-fee", st.l1Fee)
+				log.Debug("Total fee", "total-fee", mgval)
 			}
 		}
+		tonval = new(big.Int).Mul(mgval, st.tonPriceRatio)
+		// fee token is TON
+		if st.isFeeTokenUpdate {
+			// TON balance check
+			if st.state.GetTonBalance(st.msg.From()).Cmp(tonval) < 0 {
+				return errInsufficientTonBalanceForGas
+			}
+			st.state.SubTonBalance(st.msg.From(), tonval)
+			// fee token is ETH
+		} else {
+			if st.state.GetBalance(st.msg.From()).Cmp(mgval) < 0 {
+				return errInsufficientBalanceForGas
+			}
+			st.state.SubBalance(st.msg.From(), mgval)
+		}
 	}
-	if st.state.GetBalance(st.msg.From()).Cmp(mgval) < 0 {
-		return errInsufficientBalanceForGas
-	}
+	// deduct st.msg.gasLimit in the gas pool
 	if err := st.gp.SubGas(st.msg.Gas()); err != nil {
 		return err
 	}
 	st.gas += st.msg.Gas()
-
 	st.initialGas = st.msg.Gas()
```

Comment:
I think this code have to place before st.state.SubBalance. Maybe I think there is a state back point or refund logic, but we don't have to make risk.
Imagine st.state.SubBalance is successful, then gp.SubGas is error(gas limit error). State will be changed, tx will be failed.

Reply:
I think so. This is a very helpful and valuable comment. Let me change the code as you said. Thanks.